README.md
proxysocket
===========
Portable C library for proxy client support, designed to be used as a drop-in replacement for connect().

Description
-----------
Cross-platform C library to establish TCP connections using a proxy.

Supports different connection methods:
 - no proxy (optionally allowing to bind to a local address and/or port)
 - HTTP proxy: only CONNECT method, only without authentication or basic authentication
 - SOCKS4/SOCKS4A: without IDENT functionality
 - SOCKS5 (RFC 1928): only username/password authentication or no authentication

Features:
 - Currently only support IPv4 TCP connections.
 - Returns a standard operating system SOCKET that can be manipulated by standard operating system functions like send() and recv().
 - Option to perform name lookups on the proxy server.
 - Supports daisy-chaining multiple proxies.
 - Can also be used for direct connections (without proxy) optionally binding to a local address and/or port.
 - Includes a demo program that uses ipify.org to detect the public IP address being used.
 - Portable across different platforms (tested on Windows, Linux, macOS).
 
Goals
-----
Portability across platfoms (including Windows, Linux and macOS).

The returned SOCKET is a standard operating system connection handle as returned by socket(), allowing for an easy replacement of socket() and connect().

Support for IPv4 TCP connections only. Also, besides the supported proxy protocols, no specific protocol support (like HTTP or SSL) is included.

Dependancies
------------
None

License
-------
proxysocket is released under the terms of the MIT License (MIT), see LICENSE.txt.



Changelog.txt
0.1.12

2023-01-11  Brecht Sanders  https://github.com/brechtsanders/

  * added GitHub Actions CI

0.1.11

2023-01-11  Brecht Sanders  https://github.com/brechtsanders/

  * fixed Makefile to support overriding compliler with CC=<compiler>

0.1.10

2020-10-07  Brecht Sanders  https://github.com/brechtsanders/

  * fixed Makefile to detect if asprintf() is defined on Windows (needed for recent versions of MinGW-w64)

0.1.9

2020-03-07  Brecht Sanders  https://github.com/brechtsanders/

  * added Travis CI configuration file: .travis.yml
  * fixed superfluous parameter to proxysocketconfig_create_direct() in proxysocket.c
  
0.1.8

2019-09-16  Brecht Sanders  https://github.com/brechtsanders/

  * added socket_get_error_message()
  * added proxysocketconfig_set_timeout()
  * added socket_set_timeouts_milliseconds()

0.1.7

2018-10-26  Brecht Sanders  https://github.com/brechtsanders/

  * improved ipify example application check for and report errors

0.1.6

2017-10-27  Brecht Sanders  https://github.com/brechtsanders/

  * fixed issue: proxysocket_connect(): returned no error message if no logging was set

0.1.5

2017-10-20  Brecht Sanders  https://github.com/brechtsanders/

  * fixed comment for proxysocketconfig_log_fn in proxysocket.h
  * added proxysocketconfig_get_description() to get proxy description

0.1.4

2017-04-17  Brecht Sanders  https://github.com/brechtsanders/

  * proxysocket_connect() with proxy set to NULL makes direct connection
  * added proxysocketconfig_get_type_name() to get name of proxy type
  * added proxysocketconfig_get_name_type() to get proxy type by name

0.1.3

2017-03-24  Brecht Sanders  https://github.com/brechtsanders/

  * added support for proxy server host lookups (previously only local DNS)
  * example program: ipify (connects to ipify.org to determine external IP)

0.1.2

2017-03-01  Brecht Sanders  https://github.com/brechtsanders/

  * fixed memory leak

0.1.1

2016-12-17  Brecht Sanders  https://github.com/brechtsanders/

  * added function socket_receiveline()

2016-12-17  Brecht Sanders  https://github.com/brechtsanders/

  * added example application example_ipify

2016-12-05  Brecht Sanders  https://github.com/brechtsanders/

  * improved log messages

0.1.0

2016-12-05  Brecht Sanders  https://github.com/brechtsanders/

  * initial version
  * supported connection methods:
      - no proxy
      - HTTP proxy: only CONNECT method and without authentication or basic authentication
      - SOCKS5 (RFC 1928): only without authentication or username/password
      - SOCKS4: no active support for IDENT authentication


src/proxysocket_test.c
#include "proxysocket.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

//#define DST_HOST "www.edustria.be"
#define DST_HOST "www.microsoft.com"
#define DST_PORT 80
//#define DST_HOST "10.0.0.1"
//#define DST_PORT 8080
//#define DST_HOST "garbo.uwasa.fi"
//#define DST_PORT 21

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_NONE
#define PROXY_HOST NULL
#define PROXY_PORT 0
#define PROXY_USER NULL
#define PROXY_PASS NULL
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_WEB_CONNECT
#define PROXY_HOST "10.0.0.10"
#define PROXY_PORT 8080
#define PROXY_USER NULL
#define PROXY_PASS NULL
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_WEB_CONNECT
#define PROXY_HOST "localhost"
#define PROXY_PORT 3128
#define PROXY_USER "3APA3A"
#define PROXY_PASS "3apa3a"
// $MINGWPREFIX/3proxy/3proxy.exe $MINGWPREFIX/3proxy/cfg/3proxy.cfg.sample
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_SOCKS4
#define PROXY_HOST "localhost"
#define PROXY_PORT 1080
#define PROXY_USER "3APA3A"
#define PROXY_PASS "3apa3a" //ignored
// in a connected PuTTY SSH session go to Connection / SSH / Tunnels and add a forwarded port with Source port 1080, leave Destination port empty, select Dynamic and Audo and press Add
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_SOCKS5
#define PROXY_HOST "localhost"
//#define PROXY_PORT 1080
#define PROXY_PORT 1188
#define PROXY_USER NULL
#define PROXY_PASS NULL
// in a connected PuTTY SSH session go to Connection / SSH / Tunnels and add a forwarded port with Source port 1080, leave Destination port empty, select Dynamic and Audo and press Add
/**/

/**/
#define PROXY_TYPE PROXYSOCKET_TYPE_SOCKS5
#define PROXY_HOST "localhost"
#define PROXY_PORT 1080
#define PROXY_USER "3APA3A"
#define PROXY_PASS "3apa3a"
// $MINGWPREFIX/3proxy/3proxy.exe $MINGWPREFIX/3proxy/cfg/3proxy.cfg.sample
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_SOCKS4
#define PROXY_HOST "10.0.0.5"
#define PROXY_PORT 1080
#define PROXY_USER NULL
#define PROXY_PASS NULL
// ssh -D 10.0.0.1:1080 -p 22 user@server
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_SOCKS5
#define PROXY_HOST "127.0.0.1"
#define PROXY_PORT 9150
#define PROXY_USER NULL
#define PROXY_PASS NULL
// when running Tor
/**/


void logger (int level, const char* message, void* userdata)
{
  const char* lvl;
  switch (level) {
    case PROXYSOCKET_LOG_ERROR   : lvl = "ERR"; break;
    case PROXYSOCKET_LOG_WARNING : lvl = "WRN"; break;
    case PROXYSOCKET_LOG_INFO    : lvl = "INF"; break;
    case PROXYSOCKET_LOG_DEBUG   : lvl = "DBG"; break;
    default                      : lvl = "???"; break;
  }
  fprintf((FILE*)userdata, "%s: %s\n", lvl, message);
}

int main ()
{
  SOCKET sock;
  char* errmsg;
  proxysocket_initialize();
  //proxysocketconfig proxy = proxysocketconfig_create(PROXY_TYPE, PROXY_HOST, PROXY_PORT, PROXY_USER, PROXY_PASS);
//proxysocketconfig_add_proxy(proxy, PROXY_TYPE, PROXY_HOST, PROXY_PORT, PROXY_USER, PROXY_PASS);
//proxysocketconfig_add_proxy(proxy, PROXY_TYPE, PROXY_HOST, PROXY_PORT, PROXY_USER, PROXY_PASS);
//proxysocketconfig_add_proxy(proxy, PROXYSOCKET_TYPE_SOCKS4, "127.0.0.1", 1080, NULL, NULL);
//proxysocketconfig_add_proxy(proxy, PROXYSOCKET_TYPE_SOCKS5, "127.0.0.1", 1080, NULL, NULL);
proxysocketconfig proxy = proxysocketconfig_create_direct();
proxysocketconfig_add_proxy(proxy, PROXYSOCKET_TYPE_SOCKS5, "127.0.0.1", 1080, "3APA3A", "3apa3a");
proxysocketconfig_add_proxy(proxy, PROXYSOCKET_TYPE_WEB_CONNECT, "10.0.0.1", 8080, NULL, NULL);
//proxysocketconfig_add_proxy(proxy, PROXYSOCKET_TYPE_WEB_CONNECT, "10.0.0.1", 8080, NULL, NULL);
  proxysocketconfig_set_logging(proxy, logger, (void*)stdout);
  //connect
  errmsg = NULL;
  sock = proxysocket_connect(proxy, DST_HOST, DST_PORT, &errmsg);
  if (sock == INVALID_SOCKET) {
    fprintf(stderr, "%s\n", (errmsg ? errmsg : "Unknown error"));
    free(errmsg);
  } else {
    printf("----------------------------------------\n");
    //send data
    const char* http_request = "GET / HTTP/1.0\r\nHost: " DST_HOST "\r\n\r\n";
    send(sock, http_request, strlen(http_request), 0);
    //receive data
    char c;
    while (recv(sock, &c, 1, 0) > 0) {
      putchar(c);
    }
    //disconnect
    proxysocket_disconnect(proxy, sock);
  }
  proxysocketconfig_free(proxy);
  return 0;
}


//https://tools.ietf.org/html/rfc1928
//https://tools.ietf.org/rfc/rfc1929.txt
//https://tools.ietf.org/rfc/rfc1961.txt

//SOCKS4: http://ftp.icm.edu.pl/packages/socks/socks4/SOCKS4.protocol



examples/ipify.c
#include "proxysocket.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#if defined(_WIN32) && !defined(__MINGW64_VERSION_MAJOR)
#define strcasecmp stricmp
#define strncasecmp strnicmp
#endif

#define DST_HOST "api.ipify.org"
#define DST_PATH "/"
#define DST_PORT 80

void logger (int level, const char* message, void* userdata)
{
  const char* lvl;
  if (level > *(int*)userdata)
    return;
  switch (level) {
    case PROXYSOCKET_LOG_ERROR   : lvl = "ERR"; break;
    case PROXYSOCKET_LOG_WARNING : lvl = "WRN"; break;
    case PROXYSOCKET_LOG_INFO    : lvl = "INF"; break;
    case PROXYSOCKET_LOG_DEBUG   : lvl = "DBG"; break;
    default                      : lvl = "???"; break;
  }
  fprintf(stdout, "%s: %s\n", lvl, message);
}

void show_help ()
{
  printf(
    "Usage:  example_ipify [-h] [-t proxy_type] [-s proxy_server] [-p proxy_port] [-l proxy_user] [-w proxy_pass]\n"
    "Parameters:\n"
    "  -h             \tdisplay command line help\n"
    "  -t proxy_type  \ttype of proxy to use (NONE/SOCKS4/SOCKS5/WEB)\n"
    "  -s proxy_server\tproxy server host name or IP address\n"
    "  -p proxy_port  \tproxy port number\n"
    "  -l proxy_user  \tproxy authentication login\n"
    "  -w proxy_pass  \tproxy authentication password\n"
    "  -n             \tuse proxy name resolution (instead of local DNS)\n"
    "  -v             \tverbose mode\n"
    "  -d             \tdebug mode (overrides -v)\n"
    "Version: %s\n"
    "Description:\n"
    "Gets public IP address from " DST_HOST " optionally using a proxy server.\n"
    "This can be used to check if web (HTTP) access is working.", proxysocket_get_version_string()
  );
}

#define GET_PARAM()                     \
  if (argv[i][2])                       \
    param = argv[i] + 2;                \
  else if (i + 1 < argc && argv[i + 1]) \
    param = argv[++i];                  \
  else                                  \
    param = NULL;

int main (int argc, char* argv[])
{
  //get command line parameters
  int i;
  char* param;
  int proxytype = PROXYSOCKET_TYPE_NONE;
  const char* proxyhost = NULL;
  uint16_t proxyport = 0;
  const char* proxyuser = NULL;
  const char* proxypass = NULL;
  int verbose = -1;
  int proxydns = 0;
  for (i = 1; i < argc; i++) {
    //check for command line parameters
    if (argv[i][0] && (argv[i][0] == '/' || argv[i][0] == '-')) {
      switch (tolower(argv[i][1])) {
        case 'h' :
        case '?' :
          show_help();
          return 0;
        case 't' :
          GET_PARAM()
          if ((proxytype = proxysocketconfig_get_name_type(param)) == PROXYSOCKET_TYPE_INVALID) {
            fprintf(stderr, "Invalid proxy type: %s\n", param);
            show_help();
            return 1;
          }
          break;
        case 's' :
          GET_PARAM()
          if (param)
            proxyhost = param;
          break;
        case 'p' :
          GET_PARAM()
          if (param)
            proxyport = strtol(param, (char**)NULL, 10);
          break;
        case 'l' :
          GET_PARAM()
          if (param)
            proxyuser = param;
          break;
        case 'w' :
          GET_PARAM()
          if (param)
            proxypass = param;
          break;
        case 'n' :
          proxydns = 1;
          break;
        case 'v' :
          if (verbose < PROXYSOCKET_LOG_INFO)
            verbose = PROXYSOCKET_LOG_INFO;
          break;
        case 'd' :
          verbose = PROXYSOCKET_LOG_DEBUG;
          break;
        default:
          fprintf(stderr, "Invalid command line parameter: %s\n", argv[i]);
          show_help();
          return 1;
      }
    }
  }

  //make the connection via the specified proxy
  SOCKET sock;
  char* errmsg;
  //prepare for connection
  proxysocket_initialize();
  proxysocketconfig proxy = proxysocketconfig_create_direct(5);
  if (verbose >= 0)
    proxysocketconfig_set_logging(proxy, logger, (int*)&verbose);
  if (proxydns)
    proxysocketconfig_use_proxy_dns(proxy, 1);
  proxysocketconfig_add_proxy(proxy, proxytype, proxyhost, proxyport, proxyuser, proxypass);
  //connect
  errmsg = NULL;
  sock = proxysocket_connect(proxy, DST_HOST, DST_PORT, &errmsg);
  if (sock == INVALID_SOCKET) {
    fprintf(stderr, "%s\n", (errmsg ? errmsg : "Unknown error"));
  } else {
    //send data
    const char* http_request = "GET " DST_PATH " HTTP/1.0\r\nHost: " DST_HOST "\r\n\r\n";
    send(sock, http_request, strlen(http_request), 0);
    //receive data and skip header
    char* line;
    int prevempty = 0;
    int istext = 0;
    errmsg = NULL;
    while ((line = socket_receiveline(sock)) != NULL) {
      if (prevempty)
        break;
      if (strncasecmp(line, "Content-Type: text/plain", 24) == 0)
        istext = 1;
      prevempty = (line[0] ? 0 : 1);
      free(line);
    }
    if (!line) {
      errmsg = strdup("No content received");
    } else if (!istext) {
      errmsg = strdup("No plain text content returned");
    } else if (line) {
      printf("Your IP address: %s\n", line);
      free(line);
    }
    //disconnect
    proxysocket_disconnect(proxy, sock);
  }
  proxysocketconfig_free(proxy);
  if (errmsg) {
    fprintf(stderr, "Error: %s\n", errmsg);
    free(errmsg);
    return 1;
  }
  return 0;
}


src/proxysocket.h
/**
 * @file proxysocket.h
 * @brief proxysocket header file for creating network socket connections via different proxy methods
 * @author Brecht Sanders
 * @date 2016
 * @copyright MIT
 *
 * Include this header file to use proxysocket and link with -lproxysocket.
 * This header provides the functionality to connect network sockets
 * optionally via different proxy methods.
 */

#ifndef __INCLUDED_PROXYSOCKET_H
#define __INCLUDED_PROXYSOCKET_H

#include <stdint.h>

#ifdef __cplusplus
extern "C" {
#endif

/*! \cond PRIVATE */
#ifdef __WIN32__
# include <winsock2.h>
# ifdef BUILD_PROXYSOCKET_DLL
#  define DLL_EXPORT_PROXYSOCKET __declspec(dllexport)
# elif !defined(STATIC) && !defined(BUILD_PROXYSOCKET_STATIC)
#  define DLL_EXPORT_PROXYSOCKET __declspec(dllimport)
# else
#  define DLL_EXPORT_PROXYSOCKET
# endif
#else
# define DLL_EXPORT_PROXYSOCKET
# include <netinet/in.h>
# define SOCKET int
# ifndef INVALID_SOCKET
#  define INVALID_SOCKET -1
# endif
#endif
/*! \endcond */

/*! \brief version number constants
 * \sa     proxysocket_get_version()
 * \sa     proxysocket_get_version_string()
 * \name   PROXYSOCKET_VERSION_*
 * \{
 */
/*! \brief major version number */
#define PROXYSOCKET_VERSION_MAJOR 0
/*! \brief minor version number */
#define PROXYSOCKET_VERSION_MINOR 1
/*! \brief micro version number */
#define PROXYSOCKET_VERSION_MICRO 12
/*! @} */

/*! \brief proxy types
 * \sa     proxysocketconfig_create()
 * \sa     proxysocketconfig_add_proxy()
 * \{
 */
/*! \brief direct connection without proxy */
#define PROXYSOCKET_TYPE_NONE           0x00
/*! \brief SOCKS4 proxy */
#define PROXYSOCKET_TYPE_SOCKS4         0x04
/*! \brief SOCKS5 proxy */
#define PROXYSOCKET_TYPE_SOCKS5         0x05
/*! \brief HTTP proxy */
#define PROXYSOCKET_TYPE_WEB_CONNECT    0x20
/*! \brief invalid proxy type */
#define PROXYSOCKET_TYPE_INVALID        -1
/*! @} */

/*! \brief logging levels
 * \sa     proxysocketconfig_log_fn()
 * \name   PROXYSOCKET_VERSION_*
 * \{
 */
#define PROXYSOCKET_LOG_ERROR   0
#define PROXYSOCKET_LOG_WARNING 1
#define PROXYSOCKET_LOG_INFO    2
#define PROXYSOCKET_LOG_DEBUG   3
/*! @} */

//goal: function to create a socket that can be used by system function and connect it to a remote host, optionally through a proxy

/*! \brief get proxysocket version
 * \param  pmajor        pointer to integer that will receive major version number
 * \param  pminor        pointer to integer that will receive minor version number
 * \param  pmicro        pointer to integer that will receive micro version number
 * \sa     proxysocket_get_version_string()
 */
DLL_EXPORT_PROXYSOCKET void proxysocket_get_version (int* pmajor, int* pminor, int* pmicro);

/*! \brief get proxysocket version string
 * \return version string
 * \sa     proxysocket_get_version()
 */
DLL_EXPORT_PROXYSOCKET const char* proxysocket_get_version_string ();

/*! \brief get the textual name of a proxy type
 * \param  proxytype       proxy type (one of the PROXYSOCKET_TYPE_ constants)
 * \return name of the proxy type (NONE/SOCKS4A/SOCKS5/WEB/INVALID)
 * \sa     proxysocketconfig_create()
 * \sa     proxysocketconfig_get_name_type()
 */
DLL_EXPORT_PROXYSOCKET const char* proxysocketconfig_get_type_name (int proxytype);

/*! \brief get the textual name of a proxy type
 * \param  proxytypename   name of the proxy type (NONE/DIRECT/SOCKS4/SOCKS4A/SOCKS5/WEB/HTTP)
 * \return proxy type (one of the PROXYSOCKET_TYPE_ constants) or PROXYSOCKET_TYPE_INVALID on failure
 * \sa     proxysocketconfig_create()
 * \sa     proxysocketconfig_get_type_name()
 */
DLL_EXPORT_PROXYSOCKET int proxysocketconfig_get_name_type (const char* proxytypename);

/*! \brief initialize library, call once at the beginning of the main thread of the application
 * \return zero on success
 */
DLL_EXPORT_PROXYSOCKET int proxysocket_initialize ();

/*! \brief proxysocketconfig object type */
typedef struct proxysocketconfig_struct* proxysocketconfig;

/*! \brief create proxy information data structure set up for direct connection
 * \return proxy information data structure on success or NULL on failure
 * \sa     proxysocketconfig_create()
 * \sa     proxysocketconfig_add_proxy()
 * \sa     proxysocketconfig_free()
 */
DLL_EXPORT_PROXYSOCKET proxysocketconfig proxysocketconfig_create_direct ();

/*! \brief create proxy information data structure
 * \param  proxytype   proxy type (one of the PROXYSOCKET_TYPE_ constants)
 * \param  proxyhost   proxy hostname or IP address (or when proxytype is PROXYSOCKET_TYPE_NONE address to bind to if non-zero)
 * \param  proxyport   proxy port number (or when proxytype is PROXYSOCKET_TYPE_NONE port to bind to if non-zero)
 * \param  proxyuser   proxy authentication login or NULL for none
 * \param  proxypass   proxy authentication password or NULL for none
 * \return proxy information data structure on success or NULL on failure
 * \sa     proxysocketconfig_create_direct()
 * \sa     proxysocketconfig_add_proxy()
 * \sa     proxysocketconfig_free()
 */
DLL_EXPORT_PROXYSOCKET proxysocketconfig proxysocketconfig_create (int proxytype, const char* proxyhost, uint16_t proxyport, const char* proxyuser, const char* proxypass);

/*! \brief add a proxy method to proxy information data structure (multiple can be daisy chained)
 * \param  proxy       proxy information as returned by proxysocketconfig_create()
 * \param  proxytype   proxy type (one of the PROXYSOCKET_TYPE_ constants)
 * \param  proxyhost   proxy hostname or IP address (or when proxytype is PROXYSOCKET_TYPE_NONE address to bind to if non-zero)
 * \param  proxyport   proxy port number (or when proxytype is PROXYSOCKET_TYPE_NONE port to bind to if non-zero)
 * \param  proxyuser   proxy authentication login or NULL for none
 * \param  proxypass   proxy authentication password or NULL for none
 * \return proxy information data structure on success or NULL on failure
 * \sa     proxysocketconfig_create_direct()
 * \sa     proxysocketconfig_create()
 * \sa     proxysocketconfig_free()
 */
DLL_EXPORT_PROXYSOCKET int proxysocketconfig_add_proxy (proxysocketconfig proxy, int proxytype, const char* proxyhost, uint16_t proxyport, const char* proxyuser, const char* proxypass);

/*! \brief type of pointer to function for logging
 * \param  level       logging level (one of the PROXYSOCKET_LOG_ constants)
 * \param  message     text to be logged
 * \param  userdata    custom data as passed to proxysocketconfig_set_logging
 * \sa     proxysocketconfig_set_logging()
 */
typedef void (*proxysocketconfig_log_fn)(int level, const char* message, void* userdata);

/*! \brief get description of configured proxy
 * \param  proxy       proxy information as returned by proxysocketconfig_create()
 * \return string with description of proxy settings (caller must free) or NULL on failure
 */
DLL_EXPORT_PROXYSOCKET char* proxysocketconfig_get_description (proxysocketconfig proxy);

/*! \brief configure logging function
 * \param  proxy       proxy information as returned by proxysocketconfig_create()
 * \param  log_fn      logging function to use or NULL to disable logging
 * \param  userdata    user defined data that will be passed to the logging function
 * \sa     proxysocketconfig_create()
 * \sa     proxysocketconfig_log_fn
 */
DLL_EXPORT_PROXYSOCKET void proxysocketconfig_set_logging (proxysocketconfig proxy, proxysocketconfig_log_fn log_fn, void* userdata);

/*! \brief configure connection timeouts
 * \param  proxy       proxy information as returned by proxysocketconfig_create()
 * \param  sendtimeout send timeout in milliseconds (0 for no timeout)
 * \param  recvtimeout send timeout in milliseconds (0 for no timeout)
 * \sa     proxysocketconfig_create()
 */
DLL_EXPORT_PROXYSOCKET void proxysocketconfig_set_timeout (proxysocketconfig proxy, uint32_t sendtimeout, uint32_t recvtimeout);

/*! \brief specify where name resolution is done
 * \param  proxy       proxy information as returned by proxysocketconfig_create()
 * \param  proxy_dns   perform DNS lookup on the proxy server if non-zero or on client if zero (default)
 * \sa     proxysocketconfig_create()
 */
DLL_EXPORT_PROXYSOCKET void proxysocketconfig_use_proxy_dns (proxysocketconfig proxy, int proxy_dns);

/*! \brief clean up proxy information
 * \param  proxy       proxy information as returned by proxysocketconfig_create()
 * \sa     proxysocketconfig_create()
 */
DLL_EXPORT_PROXYSOCKET void proxysocketconfig_free (proxysocketconfig proxy);

/*! \brief establish a TCP connection using the specified proxy
 * \param  proxy       proxy information as returned by proxysocketconfig_create()
 * \param  dsthost     destination hostname or IP address
 * \param  dstport     destination port number
 * \param  errmsg      pointer to string that will receive error message, can be NULL, caller must free
 * \return network socket on success or INVALID_SOCKET on failure
 * \sa     proxysocketconfig_create()
 */
DLL_EXPORT_PROXYSOCKET SOCKET proxysocket_connect (proxysocketconfig proxy, const char* dsthost, uint16_t dstport, char** errmsg);

/*! \brief disconnect a proxy socket
 * \param  proxy       proxy information as returned by proxysocketconfig_create()
 * \param  sock        network socket as returned by proxysocket_connect()
 * \sa     proxysocketconfig_create()
 * \sa     proxysocket_connect()
 */
DLL_EXPORT_PROXYSOCKET void proxysocket_disconnect (proxysocketconfig proxy, SOCKET sock);

/*! \brief set socket timeout
 * \param  sock        network socket as returned by proxysocket_connect() or socket()
 * \param  timeout     timeout in seconds (0 to disable)
 * \sa     proxysocket_connect()
 */
DLL_EXPORT_PROXYSOCKET void socket_set_timeouts_milliseconds (SOCKET sock, uint32_t sendtimeout, uint32_t recvtimeout);

/*! \brief read a line from a socket
 * \param  sock        network socket as returned by proxysocket_connect() or socket()
 * \return contents of line without trailing new line (caller must free) or NULL on failure
 * \sa     proxysocket_connect()
 */
DLL_EXPORT_PROXYSOCKET char* socket_receiveline (SOCKET sock);

/*! \brief get error message of last socket error
 * \return error message or NULL if no error
 */
DLL_EXPORT_PROXYSOCKET char* socket_get_error_message ();

#ifdef __cplusplus
}
#endif

#endif //__INCLUDED_PROXYSOCKET_H


examples/example_ipify.c
#include "proxysocket.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define DST_HOST "api.ipify.org"
#define DST_PORT 80

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_NONE
#define PROXY_HOST NULL
#define PROXY_PORT 0
#define PROXY_USER NULL
#define PROXY_PASS NULL
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_WEB_CONNECT
#define PROXY_HOST "10.0.0.10"
#define PROXY_PORT 8080
#define PROXY_USER NULL
#define PROXY_PASS NULL
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_WEB_CONNECT
#define PROXY_HOST "localhost"
#define PROXY_PORT 3128
#define PROXY_USER "3APA3A"
#define PROXY_PASS "3apa3a"
// $MINGWPREFIX/3proxy/3proxy.exe $MINGWPREFIX/3proxy/cfg/3proxy.cfg.sample
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_SOCKS4
#define PROXY_HOST "localhost"
#define PROXY_PORT 1080
#define PROXY_USER "3APA3A"
#define PROXY_PASS "3apa3a" //ignored
// in a connected PuTTY SSH session go to Connection / SSH / Tunnels and add a forwarded port with Source port 1080, leave Destination port empty, select Dynamic and Audo and press Add
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_SOCKS5
#define PROXY_HOST "localhost"
//#define PROXY_PORT 1080
#define PROXY_PORT 1188
#define PROXY_USER NULL
#define PROXY_PASS NULL
// in a connected PuTTY SSH session go to Connection / SSH / Tunnels and add a forwarded port with Source port 1080, leave Destination port empty, select Dynamic and Audo and press Add
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_SOCKS5
#define PROXY_HOST "localhost"
#define PROXY_PORT 1080
#define PROXY_USER "3APA3A"
#define PROXY_PASS "3apa3a"
// $MINGWPREFIX/3proxy/3proxy.exe $MINGWPREFIX/3proxy/cfg/3proxy.cfg.sample
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_SOCKS4
#define PROXY_HOST "10.0.0.5"
#define PROXY_PORT 1080
#define PROXY_USER NULL
#define PROXY_PASS NULL
// ssh -D 10.0.0.1:1080 -p 22 user@server
/**/

/*/
#define PROXY_TYPE PROXYSOCKET_TYPE_SOCKS5
#define PROXY_HOST "127.0.0.1"
#define PROXY_PORT 9150
#define PROXY_USER NULL
#define PROXY_PASS NULL
// when running Tor
/**/

void logger (int level, const char* message, void* userdata)
{
  const char* lvl;
  if (level > *(int*)userdata)
    return;
  switch (level) {
    case PROXYSOCKET_LOG_ERROR   : lvl = "ERR"; break;
    case PROXYSOCKET_LOG_WARNING : lvl = "WRN"; break;
    case PROXYSOCKET_LOG_INFO    : lvl = "INF"; break;
    case PROXYSOCKET_LOG_DEBUG   : lvl = "DBG"; break;
    default                      : lvl = "???"; break;
  }
  fprintf(stdout, "%s: %s\n", lvl, message);
}

void show_help ()
{
  printf(
    "Usage:  example_ipify [-h] [-t proxy_type] [-s proxy_server] [-p proxy_port] [-l proxy_user] [-w proxy_pass]\n"
    "Parameters:\n"
    "  -h             \tdisplay command line help\n"
    "  -t proxy_type  \ttype of proxy to use (NONE/SOCKS4/SOCKS5/WEB)\n"
    "  -s proxy_server\tproxy server host name or IP address\n"
    "  -p proxy_port  \tproxy port number\n"
    "  -l proxy_user  \tproxy authentication login\n"
    "  -w proxy_pass  \tproxy authentication password\n"
    "  -v             \tverbose mode\n"
    "  -d             \tdebug mode\n"
    "Description:\n"
    "Gets public IP address via " DST_HOST "\n"
  );
}

int main (int argc, char* argv[])
{
  //get command line parameters
  int i;
  char* param;
  int proxytype = PROXYSOCKET_TYPE_NONE;
  const char* proxyhost = NULL;
  uint16_t proxyport = 0;
  const char* proxyuser = NULL;
  const char* proxypass = NULL;
  int verbose = -1;
  for (i = 1; i < argc; i++) {
    //check for command line parameters
    if (argv[i][0] && (argv[i][0] == '/' || argv[i][0] == '-')) {
      switch (tolower(argv[i][1])) {
        case 'h' :
        case '?' :
          show_help();
          return 0;
        case 't' :
          if (argv[i][2])
            param = argv[i] + 2;
          else if (i + 1 < argc && argv[i + 1])
            param = argv[++i];
          else
            param = NULL;
          if (stricmp(param, "SOCKS4") == 0 || stricmp(param, "SOCKS4A") == 0)
            proxytype = PROXYSOCKET_TYPE_SOCKS4;
          else if (stricmp(param, "SOCKS5") == 0)
            proxytype = PROXYSOCKET_TYPE_SOCKS5;
          else if (stricmp(param, "WEB") == 0 || stricmp(param, "HTTP") == 0)
            proxytype = PROXYSOCKET_TYPE_WEB_CONNECT;
          else {
            fprintf(stderr, "Invalid proxy type: %s\n", param);
            show_help();
            return 1;
          }
          break;
        case 's' :
          if (argv[i][2])
            param = argv[i] + 2;
          else if (i + 1 < argc && argv[i + 1])
            param = argv[++i];
          else
            param = NULL;
          if (param)
            proxyhost = param;
          break;
        case 'p' :
          if (argv[i][2])
            param = argv[i] + 2;
          else if (i + 1 < argc && argv[i + 1])
            param = argv[++i];
          else
            param = NULL;
          if (param)
            proxyport = strtol(param, (char**)NULL, 10);
          break;
        case 'l' :
          if (argv[i][2])
            param = argv[i] + 2;
          else if (i + 1 < argc && argv[i + 1])
            param = argv[++i];
          else
            param = NULL;
          if (param)
            proxyuser = param;
          break;
        case 'w' :
          if (argv[i][2])
            param = argv[i] + 2;
          else if (i + 1 < argc && argv[i + 1])
            param = argv[++i];
          else
            param = NULL;
          if (param)
            proxypass = param;
          break;
        case 'v' :
          verbose = PROXYSOCKET_LOG_INFO;
          break;
        case 'd' :
          verbose = PROXYSOCKET_LOG_DEBUG;
          break;
        default:
          fprintf(stderr, "Invalid command line parameter: %s\n", argv[i]);
          show_help();
          return 1;
      }
    }
  }

  SOCKET sock;
  char* errmsg;
  //prepare for connection
  proxysocket_initialize();
  proxysocketconfig proxy = proxysocketconfig_create_direct();
  if (verbose >= 0)
    proxysocketconfig_set_logging(proxy, logger, (int*)&verbose);
  proxysocketconfig_use_proxy_dns(proxy, 1);/////
  proxysocketconfig_add_proxy(proxy, proxytype, proxyhost, proxyport, proxyuser, proxypass);
  //connect
  errmsg = NULL;
  sock = proxysocket_connect(proxy, DST_HOST, DST_PORT, &errmsg);
  if (sock == INVALID_SOCKET) {
    fprintf(stderr, "%s\n", (errmsg ? errmsg : "Unknown error"));
    free(errmsg);
  } else {
    //send data
    const char* http_request = "GET / HTTP/1.0\r\nHost: " DST_HOST "\r\n\r\n";
    send(sock, http_request, strlen(http_request), 0);
    //receive data and skip header
    char* line;
    int prevempty = 0;
    while ((line = socket_receiveline(sock)) != NULL) {
      if (prevempty)
        break;
      prevempty = (line[0] ? 0 : 1);
      free(line);
    }
    if (line) {
      printf("Your IP address: %s\n", line);
      free(line);
    }
    //disconnect
    proxysocket_disconnect(proxy, sock);
  }
  proxysocketconfig_free(proxy);
  return 0;
}
